sql/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteThreadRunner.scala

```diff
@@ -68,10 +68,19 @@ private[connect] class ExecuteThreadRunner(executeHolder: ExecuteHolder) extends
   }
 
   /**
-   * Register a callback that gets executed after completion/interruption of the execution thread.
+   * Execute the supplied lambda if no execute runner thread has started, or pass it to the
+   * execute runner thread so that the lambda will get executed on completion. Failure details
+   * will only be passed to the lambda only when an error was raised during actual execution.
    */
-  private[connect] def processOnCompletion(callback: Try[Unit] => Unit): Unit = {
-    promise.future.onComplete(callback)(ExecuteThreadRunner.namedExecutionContext)
+  private[connect] def processNowOrOnCompletion(callback: Try[Unit] => Unit): Unit = {
+    lock.synchronized {
+      if (started) {
+        promise.future.onComplete(callback)(ExecuteThreadRunner.namedExecutionContext)
+      } else {
+        // No thread has been spawned, and therefore there was no failure, passing `Success`.
+        callback(Success(()))
```

Comment:
Does this callback have to be executed within the lock?

Reply:
HI, that's what I'm fixing; incoming new change in 1 minute.

Reply:
the whole "promise" and locking won't be needed for this as what this "promise" does is pretty static. I'll try to get rid of most of those complicated code pieces.